File: frameworks/ts/src/context.ts
/**
 * Context object — passed to every processor invocation.
 *
 * Mirrors the Python and Rust Context API:
 *   ctx.output(v, {id?, src?})
 *   ctx.emit(queue, v, {id?, src?})
 *   ctx.drain()
 *   ctx.meta(v)
 *   ctx.log(msg, {level?, ...extra})
 *   ctx.error(v, err)
 *   ctx.trace(k, v)
 *   ctx.stats(data)
 *   ctx.hash(str)
 *   ctx.hashFile(path)
 */

import type { JSONValue } from './envelope';
import {
  hashFile,
  hashString,
  writeData,
  writeError,
  writeLog,
  writeMeta,
  writeStats,
  writeTrace,
} from './envelope';

export interface QueueItem {
  queue: string;
  v: JSONValue;
  id: string;
  src: string;
}

/** Tiny runtime surface the Context needs — avoids circular imports with Runtime. */
export interface RuntimeLike {
  enqueue(item: QueueItem): void;
  drainQueue(): Promise<void>;
}

export class Context {
  readonly id: string;
  readonly src: string;
  readonly memory: Memory;
  private readonly _runtime: RuntimeLike;
  /** Labels accumulated by ctx.trace(k, v). Flushed (merged) as one
   *  {type:"trace"} stderr event before each ctx.output(), then cleared. */
  private _labels: Record<string, JSONValue> = {};

  constructor(id: string, src: string, memory: Memory, runtime: RuntimeLike) {
    this.id = id;
    this.src = src;
    this.memory = memory;
    this._runtime = runtime;
  }

  /** Accumulate a label on this invocation's next output envelope. */
  trace(key: string, value: JSONValue): void {
    this._labels[key] = value;
  }

  /** Emit a stats event to stderr. Example: ctx.stats({rows:1000, rps:250}). */
  stats(data: Record<string, JSONValue>): void {
    if (data && Object.keys(data).length > 0) {
      writeStats(data);
    }
  }

  output(v: JSONValue, opts: { id?: string; src?: string } = {}): void {
    const outId = opts.id ?? this.id;
    const outSrc = opts.src ?? this.src;
    // Emit merged trace first (always, even with empty labels), then data, then clear.
    writeTrace(outId, outSrc, this._labels);
    this._labels = {};
    writeData(v, outId, outSrc);
  }

  emit(queue: string, v: JSONValue, opts: { id?: string; src?: string } = {}): void {
    this._runtime.enqueue({
      queue,
      v,
      id: opts.id ?? this.id,
      src: opts.src ?? this.src,
    });
  }

  async drain(): Promise<void> {
    await this._runtime.drainQueue();
  }

  meta(v: JSONValue): void {
    writeMeta(v);
  }

  log(msg: string, opts: { level?: string } & Record<string, JSONValue> = {}): void {
    const { level = 'info', ...extra } = opts as { level?: string } & Record<string, JSONValue>;
    writeLog(msg, level, Object.keys(extra).length ? extra : undefined);
  }

  error(v: JSONValue, err: unknown): void {
    writeError(v, err, this.id, this.src);
  }

  hash(key: string): string {
    return hashString(key);
  }

  hashFile(filepath: string): Promise<string | undefined> {
    return hashFile(filepath);
  }
}

/**
 * Shared accumulator memory. Tools extend this by attaching their own
 * accumulator instances as properties. Named accumulator types will ship in a
 * later milestone — filter tool doesn't need them.
 */
export class Memory {
  [key: string]: unknown;
}

File: frameworks/rust/src/context.rs
//! Context object — passed to every processor invocation.

use serde_json::{Map, Value};
use std::io::Write;

use crate::accumulators::Memory;
use crate::envelope;

/// Queue item: (queue_name, v, id, src)
pub type QueueItem = (String, Value, String, String);

/// Processing context created per invocation.
pub struct Context<'a> {
    pub id: String,
    pub src: String,
    pub memory: &'a mut Memory,
    queue: &'a mut Vec<QueueItem>,
    stdout: &'a mut dyn Write,
    stderr: &'a mut dyn Write,
    /// Labels accumulated by ctx.trace(k, v). Flushed (merged) as one
    /// {type:"trace"} stderr event before each ctx.output(), then cleared.
    labels: Map<String, Value>,
}

impl<'a> Context<'a> {
    pub fn new(
        id: String,
        src: String,
        memory: &'a mut Memory,
        queue: &'a mut Vec<QueueItem>,
        stdout: &'a mut dyn Write,
        stderr: &'a mut dyn Write,
    ) -> Self {
        Self { id, src, memory, queue, stdout, stderr, labels: Map::new() }
    }

    /// Accumulate a label on this invocation's next output envelope.
    pub fn trace(&mut self, key: &str, value: Value) {
        self.labels.insert(key.to_string(), value);
    }

    /// Emit a stats event to stderr.
    pub fn stats(&mut self, data: Value) {
        if data.is_object() && !data.as_object().map(|o| o.is_empty()).unwrap_or(true) {
            envelope::write_stats(&data, &mut self.stderr);
        }
    }

    /// Emit data record to stdout.
    /// Flushes accumulated trace labels (merged) first, then the envelope.
    pub fn output(&mut self, v: Value, id: Option<&str>, src: Option<&str>) {
        let out_id = id.unwrap_or(&self.id).to_string();
        let out_src = src.unwrap_or(&self.src).to_string();
        let labels_value = Value::Object(std::mem::take(&mut self.labels));
        envelope::write_trace(&out_id, &out_src, &labels_value, &mut self.stderr);
        envelope::write_data(&v, &out_id, &out_src, &mut self.stdout);
    }

    /// Emit to internal named queue.
    pub fn emit(&mut self, queue_name: &str, v: Value, id: Option<&str>, src: Option<&str>) {
        let item_id = id.unwrap_or(&self.id).to_string();
        let item_src = src.unwrap_or(&self.src).to_string();
        self.queue.push((queue_name.to_string(), v, item_id, item_src));
    }

    /// Emit metadata record to stdout.
    pub fn meta(&mut self, v: Value) {
        envelope::write_meta(&v, &mut self.stdout);
    }

    /// Write structured log to stderr.
    pub fn log(&mut self, msg: &str, level: &str) {
        envelope::write_log(msg, level, &mut self.stderr);
    }

    /// Write error to stderr with original input preserved.
    pub fn error(&mut self, v: &Value, err: &str) {
        envelope::write_error(v, err, &self.id, &self.src, &mut self.stderr);
    }

    /// Hash a string. Returns 16-char hex.
    pub fn hash(&self, key: &str) -> String {
        envelope::hash_string(key)
    }

    /// Hash file content in chunks. Returns hex string or None on error.
    pub fn hash_file(&self, filepath: &str) -> Option<String> {
        envelope::hash_file(filepath, 65536)
    }
}
